pkg/provider/util/instancetypes/instancetypes.go

```diff
@@ -0,0 +1,104 @@
+package instancetypes
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/amazon-ec2-instance-selector/v2/pkg/bytequantity"
+	"github.com/aws/amazon-ec2-instance-selector/v2/pkg/selector"
+	"github.com/aws/aws-sdk-go-v2/config"
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+)
+
+type arch int
+
+func (a arch) String() string {
+	switch a {
+	case Amd64:
+		return "x64"
+	case Arm64:
+		return "Arm64"
+	}
+	return ""
+}
+
+const (
+	Amd64 arch = iota + 1
+	Arm64
+)
+
+const maxResults = 15 // maximum number of VM types to fetch
+
+type AwsInstanceRequest struct {
+	CPUs       int32
+	MemoryGib  int32
+	Arch       arch
+	NestedVirt bool
+}
+
+type AzureInstanceRequest struct {
+	CPUs       int32
+	MemoryGib  int32
+	Arch       arch
+	NestedVirt bool
+}
+
+func validate(cpus, memory int32, arch arch) error {
+	if cpus > 0 && memory > 0 && arch.String() != "" {
+		return nil
+	}
+	return fmt.Errorf("Invalid values for CPUs: %d, Memory: %d and Arch: %s", cpus, memory, arch)
+}
+
+func (r *AwsInstanceRequest) GetMachineTypes() ([]string, error) {
```

Comment:
May I would suggest change a bit the structure of the pkg here, I would create a interface for 

```go
func  GetMachineTypes() ([]string, error) {
```

And would be implemented by azure and aws, each specific code would be on its own file i.e:

instancesTypes, instancesTypesAzure, instancesTypesAWS, probably for the interface approach you would require an extra module on the same pkg (api) similar to https://github.com/adrianriobo/goax/tree/main/pkg/goax/app

does this make sense to you? I think it would organize the functions a bit better as right now they are kind of mixed on instancesTypes, and eventually if we add gpc support it would add more code on same file

Reply:
yeah, i can try this approach will update the PR